contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -274,6 +284,326 @@ transform_funcexpr(Node *node)
+					*concat_expr;
+		Node	   *constant_suffix;
+		Const	   *highest_sort_key;
+
+		/* construct leftop >= pattern */
+		optup = compatible_oper(NULL, list_make1(makeString(">=")), ltypeId, ltypeId,
+								true, -1);
+		if (optup == (Operator) NULL)
+			return node;
+		greater_equal = make_op_with_func(oprid(optup), BOOLOID, false,
+											(Expr *) leftop, (Expr *) prefix,
+											InvalidOid, server_collation_oid, oprfuncid(optup));
+		ReleaseSysCache(optup);
+		/* construct pattern||E'\uFFFF' */
+		highest_sort_key = makeConst(TEXTOID, -1, server_collation_oid, -1,
+										PointerGetDatum(cstring_to_text(SORT_KEY_STR)), false, false);
+
+		optup = compatible_oper(NULL, list_make1(makeString("||")), rtypeId, rtypeId,
+								true, -1);
+		if (optup == (Operator) NULL)
+			return node;
+		concat_expr = make_op_with_func(oprid(optup), rtypeId, false,
+										(Expr *) prefix, (Expr *) highest_sort_key,
+										InvalidOid, server_collation_oid, oprfuncid(optup));
+		ReleaseSysCache(optup);
+		/* construct leftop < pattern */
+		optup = compatible_oper(NULL, list_make1(makeString("<")), ltypeId, ltypeId,
+								true, -1);
+		if (optup == (Operator) NULL)
+			return node;
+
+		less_equal = make_op_with_func(oprid(optup), BOOLOID, false,
+										(Expr *) leftop, (Expr *) concat_expr,
+										InvalidOid, server_collation_oid, oprfuncid(optup));
+		constant_suffix = make_and_qual((Node *) greater_equal, (Node *) less_equal);
+		if (like_entry.is_not_match)
+		{
+			constant_suffix = (Node *) make_notclause((Expr *) constant_suffix);
+			ret = make_or_qual(node, constant_suffix);
+		}
+		else
+		{
+			constant_suffix = make_and_qual((Node *) greater_equal, (Node *) less_equal);
+			ret = make_and_qual(node, constant_suffix);
+		}
+		ReleaseSysCache(optup);
+	}
+	return ret;
+}
+
+static void
+get_remove_accents_internal_oid()
+{
+	Oid funcargtypes[1] = {TEXTOID};
+	if (remove_accents_internal_oid)
+		return;
+
+	remove_accents_internal_oid = LookupFuncName(list_make2(makeString("sys"), makeString("remove_accents_internal")), -1, funcargtypes, true);
+}
+
+/*
+ * Function responsible for obtaining unaccented version of input
+ * string with the help of ICU provided APIs. 
+ * We use a transformation rule to transliterate the string
+ */
+
+PG_FUNCTION_INFO_V1(remove_accents_internal);
+Datum
+remove_accents_internal(PG_FUNCTION_ARGS)
+{
+	char *input = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	UChar *rules; // Rule to remove diacritics 
+	UChar *utf16_intput;
+	int32_t len_uchar,
+			len_uinput,
+			limit,
+			capacity,
+			len_result;
+	UErrorCode status = U_ZERO_ERROR;
+	char *result;
+	UTransliterator *transliterator;
+
+#ifdef USE_ICU
+	len_uchar = icu_to_uchar(&rules, TRANSFORMATION_RULE, strlen(TRANSFORMATION_RULE));
+	len_uinput = icu_to_uchar(&utf16_intput, input, strlen(input));
+	
+	/* Open a transliterator for conversion */
+	transliterator = utrans_openU(rules, len_uchar, UTRANS_FORWARD, NULL, 0, NULL, &status);
```

Comment:
we should cache transliterator in TopMemoryContext to save call to utrans_openU every time

Reply:
updated